File: packages/cargo/jns42-generator/src/models/arena.rs
#[derive(Clone, Debug, PartialEq, Default)]
pub struct Arena<T>(pub(super) im::Vector<T>)
where
  T: Clone + PartialEq;

impl<T> Arena<T>
where
  T: Clone + PartialEq,
{
  pub fn new() -> Self {
    Self(im::Vector::new())
  }

  pub fn get_item(&self, key: usize) -> &T {
    self.0.get(key).unwrap()
  }

  pub fn set_item(&mut self, key: usize, item: T) -> T {
    self.0.set(key, item)
  }

  pub fn add_item(&mut self, item: T) -> usize {
    let key = self.0.len();
    self.0.push_back(item);
    key
  }

  pub fn iter(&self) -> impl Iterator<Item = &T> {
    self.0.iter()
  }

  pub fn apply_transform(&mut self, transform: impl Fn(&mut Self, usize)) -> usize {
    let mut count = 0;
    let mut key = 0;
    while key < self.0.len() {
      let self_previous = self.clone();

      transform(self, key);

      if self.0 != self_previous.0 {
        count += 1
      }

      key += 1;
    }

    count
  }
}

File: packages/npm/jns42-generator/src/utils/arena.ts
import immutable from "immutable";

export type ArenaTransform<T, A extends Arena<T> = Arena<T>> = (arena: A, key: number) => void;

export class Arena<T> {
  private items: immutable.List<T>;

  constructor(items: Iterable<T> = []) {
    this.items = immutable.List(items);
  }

  public [Symbol.iterator]() {
    return this.items[Symbol.iterator]();
  }

  public get size() {
    return this.items.count();
  }

  public getItem(key: number): T {
    return this.items.get(key)!;
  }

  public setItem(key: number, item: T) {
    this.items = this.items.set(key, item);
  }

  public addItem(item: T): number {
    const key = this.items.count();
    this.items = this.items.push(item);
    return key;
  }

  public applyTransform(...transformers: ArenaTransform<T, typeof this>[]): number {
    let counter = 0;
    // one iteration has a set number of keys to loop through, this is
    // because then we can detect infinite loops. So newly added items
    // will have to wait for the next iteration to be transformed
    const count = this.items.count();
    for (let key = 0; key < count; key++) {
      for (const transform of transformers) {
        let itemsPrevious = this.items;
        transform(this, key);
        if (this.items !== itemsPrevious) {
          counter++;
        }
      }
    }
    return counter;
  }
}
